src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -176,17 +176,22 @@ private IEnumerable<TypeInferenceInfo> GetTypesSimple(ExpressionSyntax expressio
 
             private IEnumerable<TypeInferenceInfo> InferTypeInArrowExpressionClause(ArrowExpressionClauseSyntax arrowClause)
             {
-                if (arrowClause.IsParentKind(SyntaxKind.PropertyDeclaration))
-                {
-                    return InferTypeInPropertyDeclaration(arrowClause.Parent as PropertyDeclarationSyntax);
-                }
+                var parentSymbol = SemanticModel.GetDeclaredSymbol(arrowClause.Parent);
+                var symbolReturnType = GetSymbolReturnType();
+
+                return symbolReturnType != null 
+                    ? SpecializedCollections.SingletonEnumerable(new TypeInferenceInfo(symbolReturnType))
+                    : SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
 
-                if (arrowClause.Parent is BaseMethodDeclarationSyntax)
+                ITypeSymbol GetSymbolReturnType()
                 {
-                    return InferTypeInBaseMethodDeclaration(arrowClause.Parent as BaseMethodDeclarationSyntax);
+                    switch (parentSymbol)
+                    {
+                        case IPropertySymbol property: return property?.Type;
+                        case IMethodSymbol method: return method?.ReturnType;
```

Comment:
nit: ?. is unnecessary, if `parentSymbol` is null, none of the patterns will match